tools/stylelint/theme-mixin-api.js

```diff
@@ -0,0 +1,113 @@
+const stylelint = require('stylelint');
+
+/** Name of this stylelint rule. */
+const ruleName = 'material/theme-mixin-api';
+
+/** Regular expression that matches all theme mixins. */
+const themeMixinRegex =
+    /^(?:_(mat-.+)-(density)|(mat-.+)-(density|color|typography|theme))\(([^,]+)\)$/;
+
+/**
+ * Stylelint plugin which ensures that theme mixins have a consistent API. Besides
+ * compilation API tests which are stored in `src/material/core/theming/test`, we test
+ * the following patterns here:
+ *
+ *   1. Checks if theme mixin arguments named consistently e.g. if a mixin accepts both a theme
+ *      or color configuration, the variable should reflect that.
+ *   2. Checks if the individual theme mixins handle the case where consumers pass a theme object.
+ *      For convenience, we support passing theme object to the scoped mixins.
+ *   3. Checks if the `-theme` mixins have the duplicate style check set up. We want to
+ *      consistently check for duplicative theme styles so that we can warn consumers.
+ */
+const plugin = stylelint.createPlugin(ruleName, (isEnabled, options, context) => {
+  return (root, result) => {
+    if (!isEnabled)
+      return;
+
+    root.walkAtRules('mixin', node => {
+      const matches = node.params.match(themeMixinRegex);
+      if (matches === null) {
+        return;
+      }
+
+      const componentName = matches[1] || matches[3];
+      const type = matches[2] ||matches[4];
+      const variableName = matches[5];
+
+      if (type === 'theme') {
+        validateThemeMixin(node, componentName, variableName);
+      } else {
+        validateIndividualSystemMixins(node, type, variableName);
+      }
+    });
+
+    function validateThemeMixin(node, componentName, variableName) {
+      if (variableName !== '$theme-or-color-config') {
+        if (context.fix) {
+          node.params = node.params.replace(variableName, '$theme-or-color-config');
+        } else {
+          reportError(node, 'Expected mixin argument to be called `$theme-or-color-config`.');
+        }
+      }
+
+      const themePropName = `$theme`;
```

Comment:
should we check the actual name of the variable instead of assuming $theme, if it were something else, the error probably wouldn't make sense - or add an error if the name isn't theme so the error makes sense

Reply:
Yeah that's my point though, if I do `$theme-config: _mat-legacy-get-theme(...);` I'll get a confusing lint error about not using `_mat-legacy-get-theme`, even though I am using it

Reply:
I was hesitant in doing this as it results in a more convoluted rule, but looks like it's not too much code. Added.